### DISCUSSION TITLE
Usage on the Android devices

### DISCUSSION THREAD

**Discussion by menion** [OSGeo/PROJ-JNI#69] [Q&A] [answered]
Hello,
I would like to ask for a small (hopefully) help with usage JNI binding on the Android platform.

I was able to succesfully compile PROJ 9.x + PROJ-JNI 2.0 library. 
The result is project with
- libproj.so, libproj-binging.so & libsqlite3.so libraries in the project "jniLibs" directory
- direct copy of "java" source in the library 
  - where I just had to modify loading of library in the `NativeResource`, no a problem.

**Problem**
After start and calling sample `TransformPoints` sample, I get following error.

As I'm not well skilled with C code, I'm still unable to find out why and where this happens. `java.util.logging.Logger` package IS available on the Android devices.

Appreciate and help, thanks!

```
 A  java_vm_ext.cc:594] JNI DETECTED ERROR IN APPLICATION: the return type of CallObjectMethodV does not match void java.util.logging.Logger.log(java.util.logging.Level, java.lang.String)
    java_vm_ext.cc:594]     in call to CallObjectMethodV
    java_vm_ext.cc:594]     from long org.osgeo.proj.AuthorityFactory.newInstance(org.osgeo.proj.Context, java.lang.String)
 A  runtime.cc:676] Runtime aborting...
    runtime.cc:676] Dumping all threads without mutator lock held
    runtime.cc:676] All threads:
    runtime.cc:676] DALVIK THREADS (18):
    runtime.cc:676] "main" prio=10 tid=1 Runnable
    runtime.cc:676]   | group="" sCount=0 ucsCount=0 flags=0 obj=0x71e83728 self=0xb4000076908b3f50
    runtime.cc:676]   | sysTid=19071 nice=-10 cgrp=system sched=0/0 handle=0x77d6abc4f8
    runtime.cc:676]   | state=R schedstat=( 249233600 3209686 120 ) utm=20 stm=4 core=6 HZ=100
    runtime.cc:676]   | stack=0x7fc7a84000-0x7fc7a86000 stackSize=8188KB
    runtime.cc:676]   | held mutexes= "abort lock" "mutator lock"(shared held)
    runtime.cc:676]   native: #00 pc 0000000000535c24  /apex/com.android.art/lib64/libart.so (art::DumpNativeStack(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, int, BacktraceMap*, char const*, art::ArtMethod*, void*, bool)+128) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #01 pc 00000000006efd94  /apex/com.android.art/lib64/libart.so (art::Thread::DumpStack(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, bool, BacktraceMap*, bool) const+236) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #02 pc 00000000006fd620  /apex/com.android.art/lib64/libart.so (art::DumpCheckpoint::Run(art::Thread*)+208) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #03 pc 0000000000362a60  /apex/com.android.art/lib64/libart.so (art::ThreadList::RunCheckpoint(art::Closure*, art::Closure*)+440) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #04 pc 00000000006fbdc0  /apex/com.android.art/lib64/libart.so (art::ThreadList::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, bool)+280) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #05 pc 00000000006d6e28  /apex/com.android.art/lib64/libart.so (art::AbortState::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) const+212) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #06 pc 00000000006d1b14  /apex/com.android.art/lib64/libart.so (art::Runtime::Abort(char const*)+1016) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #07 pc 0000000000016ea8  /apex/com.android.art/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_3::__invoke(char const*)+80) (BuildId: 420d56eac27a210c92900f3ddb760c86)
    runtime.cc:676]   native: #08 pc 0000000000016450  /apex/com.android.art/lib64/libbase.so (android::base::LogMessage::~LogMessage()+352) (BuildId: 420d56eac27a210c92900f3ddb760c86)
    runtime.cc:676]   native: #09 pc 0000000000442e24  /apex/com.android.art/lib64/libart.so (art::JavaVMExt::JniAbort(char const*, char const*)+1612) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #10 pc 000000000032b1d0  /apex/com.android.art/lib64/libart.so (art::JavaVMExt::JniAbortV(char const*, char const*, std::__va_list)+108) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #11 pc 000000000048967c  /apex/com.android.art/lib64/libart.so (art::(anonymous namespace)::ScopedCheck::AbortF(char const*, ...) (.__uniq.99033978352804627313491551960229047428)+144) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #12 pc 0000000000473f20  /apex/com.android.art/lib64/libart.so (art::(anonymous namespace)::ScopedCheck::CheckMethodAndSig(art::ScopedObjectAccess&, _jobject*, _jclass*, _jmethodID*, art::Primitive::Type, art::InvokeType) (.__uniq.99033978352804627313491551960229047428)+1032) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #13 pc 0000000000472970  /apex/com.android.art/lib64/libart.so (art::(anonymous namespace)::CheckJNI::CallMethodV(char const*, _JNIEnv*, _jobject*, _jclass*, _jmethodID*, std::__va_list, art::Primitive::Type, art::InvokeType) (.__uniq.99033978352804627313491551960229047428)+304) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
 A  runtime.cc:676]   native: #14 pc 0000000000472818  /apex/com.android.art/lib64/libart.so (art::(anonymous namespace)::CheckJNI::CallObjectMethodV(_JNIEnv*, _jobject*, _jmethodID*, std::__va_list) (.__uniq.99033978352804627313491551960229047428.llvm.13572457048004851375)+72) (BuildId: 28c5aa8a2e8fc5df069f717d6e94f7fe)
    runtime.cc:676]   native: #15 pc 0000000000014bc4  /data/app/~~1BG_NeZ11vyjXDO3_0YS4g==/com.asamm.locus.gdal.sample-xwFf_AsYVNLlCBoIMdBQZA==/lib/arm64/libproj-binding.so (_JNIEnv::CallObjectMethod(_jobject*, _jmethodID*, ...)+116) (BuildId: 31a2e685b5d8a5c1ff8f3166ef450d69342570b3)
    runtime.cc:676]   native: #16 pc 0000000000014a5c  /data/app/~~1BG_NeZ11vyjXDO3_0YS4g==/com.asamm.locus.gdal.sample-xwFf_AsYVNLlCBoIMdBQZA==/lib/arm64/libproj-binding.so (log(_JNIEnv*, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&)+240) (BuildId: 31a2e685b5d8a5c1ff8f3166ef450d69342570b3)
    runtime.cc:676]   native: #17 pc 0000000000015c0c  /data/app/~~1BG_NeZ11vyjXDO3_0YS4g==/com.asamm.locus.gdal.sample-xwFf_AsYVNLlCBoIMdBQZA==/lib/arm64/libproj-binding.so (get_database_context(_JNIEnv*, _jobject*)+292) (BuildId: 31a2e685b5d8a5c1ff8f3166ef450d69342570b3)
    runtime.cc:676]   native: #18 pc 0000000000020da4  /data/app/~~1BG_NeZ11vyjXDO3_0YS4g==/com.asamm.locus.gdal.sample-xwFf_AsYVNLlCBoIMdBQZA==/lib/arm64/libproj-binding.so (Java_org_osgeo_proj_AuthorityFactory_newInstance+204) (BuildId: 31a2e685b5d8a5c1ff8f3166ef450d69342570b3)
... more
```

**Comment by desruisseaux** [comment, marked as answer]
I do not really know, since I have few experience in Android development. The [Android API documentation](https://developer.android.com/reference/java/util/logging/Logger#log(java.util.logging.Level,%20java.lang.String)) suggests that this method exists and has the expected signature.

There is a test that we could do. In the `bindings.cpp` file, around lines 248 to 260 there is this code:

```java
jclass c = env->FindClass("java/util/logging/Logger");
if (c) {
    java_method_log = env->GetMethodID(c, "log", "(Ljava/util/logging/Level;Ljava/lang/String;)V");
    if (java_method_log) {
        c = env->FindClass("java/util/logging/Level");
        if (c) {
            java_field_debug_level = env->GetStaticFieldID(c, "FINE", "Ljava/util/logging/Level;");
            if (java_field_debug_level) {
                java_method_getLogger = env->GetStaticMethodID(caller, "logger", "()Ljava/util/logging/Logger;");
            }
        }
    }
}
```

Try to add another `env->FindClass("java/util/logging/Level");` call just before the first `GetMethodID(…)`. This is for testing if it causes Android to perform some kind of class resolution before we use `Level` in method signature. If it works, I would like to know in order to modify the C++ code in a cleaner way.

If it doesn't work, you can try to comment-out the whole above block. PROJ-JNI should still work with only logging disabled.

**Comment by menion** [comment]
Amazing, thanks!

You pointed to the correct place. When this block is commented, it works as expected (except that the PROJ has the problem to find proj.db, but this is a solvable problem).
Seems that the "log" library is not included in the build. So I'm trying to add it manually by modifying `CMakeLists.txt` but no success so far. Anyway, this is not a serious problem. 

Thanks again for the help, I'm moving forward.

**Comment by desruisseaux** [comment]
Alternatively I pushed a commit where I try to change the initialization order. This is only a guess, I have not idea if it has any incidence for Android. If you have a chance, would it be possible to test?

Commit: https://github.com/OSGeo/PROJ-JNI/commit/6be80ee304b313ca28a9f4b868d279e942b645ac

**Comment by menion** [comment]
Hello, thanks, appreciate your help. Changes you made do not help here anyway, same error.

I believe that the problem is in a different place. 

In Android, native libs are usually built by Android.mk where logger may be added by `LOCAL_LDLIBS += -llog`. What I most probably miss is the same, but over `cmake`. So with my limited knowledge, it should be necessary to modify the CMakeLists.txt file and `add_library` somehow. But all the tries I did were not working.

Anyway, as I wrote, with commenting problematic block, all seems to work for now and this is a sufficient solution for me for now. Thanks!

**Comment by Hinute** [comment]
> I was able to succesfully compile PROJ 9.x + PROJ-JNI 2.0 library. The result is project with
> 
> * libproj.so, libproj-binging.so & libsqlite3.so libraries in the project "jniLibs" directory
> * direct copy of "java" source in the library
>   
>   * where I just had to modify loading of library in the `NativeResource`, no a problem.
> 
@menion would you be able to give any details about how you modified NativeResource to properly load the library? I am also trying to get PROJ JNI to compile for Android and am running into the issue where it fails to find my libproj-binding.so file.

**Comment by akkie2106** [comment]
how to implement this lib in the android app? Please give steps

**Reply by desruisseaux** [reply]
Hello. As fair as I know, the current maintainers of PROJ-JNI are not familiar with Android development. A step-by-step guide would need to be written by other volunteers, maybe on a wiki page if someone is willing to take this task.

**Reply by JigneshPiZone** [reply]
@desruisseaux Any update on this?

**Reply by desruisseaux** [reply]
No work that I'm aware of. Do you have the same issue as described above, i.e. `java.util.logging` not found? If yes, does the [above workaround](https://github.com/OSGeo/PROJ-JNI/discussions/69#discussioncomment-5188128) works?
